#5072 Implement contextual feedback modal on data file submission

lhawkman27 opened the pull request:
## Summary of Changes

- Created a "Feedback Widget/Lightweight Modal" component that mirrors the look of the "Feedback Modal" and feedback submission logic
- Added functionality to open the Feedback widget on successful Data Files upload whether TANF or FRA data files
- Header for the widget changes based on the specific file you are submitting
- Widget doesn't expend until user selects a rating (lightweight validation for required selection)
- Once rating is selected selection specific text shows above text area 
- Once user submits feedback, the Feedback Widget's text changes to saying "Thank you for your Feedback" with a SPINNER that animates for 5 secs and then auto closes the widget. (Users can click on the close button if they do not want to wait the 5 secs)
- Created a useFocusTrap hook so tabbing and keydown logic can be shared between various components
- Added in logic for widget's focus trap to take affect once user tabs into it or user specifically clicks on the widget
- Once focus on the widget users can click off of it to get out of it's focus trap or hit ESC to close the widget (aligns with accessibility norms)
- Connected the frontend with the backend Feedback API  endpoint

Pull request closes #5072 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

Test General Feedback Modal Feedback submission (Backend integration)
1. Open http://localhost:3000/ and sign in
2. Click the "Give Feedback" button in the bottom right corner.
3. Select a User Experience rating. (Users can click or tab to selection and hit ENTER)
4. Enter in "Tell Us More" text if you want
5. Click "Send Feedback"
6. See "Thank you for you feedback" message and close it
7. Navigate to Admin tab 
8. Scroll down to Users section and select Feedback
9. Confirm you see entires in the Feedback View
------------------------------------------------------------------------------
Feedback Widget UI and Feedback Submission Test

1. Open http://localhost:3000/ and sign in
2. Navigate to FRA Data Files
3. Enter in any FRA Data Reporting Information that you want and Search
4. Upload a proper data file
5. Should see "Successfully submitted alert" and the "Feedback Widget" should appear in the bottom right above the website footer
6. Click on the widget and it should be focused on now 
7. Hit the Tab key and cycle through the visible elements
8. Click outside of the Widget and hit tab again
9. Confirm Tab is not focused in the Widget
10. Now click the widget again
11. Selecting a rating will expand the widget to it's full form (Users can click or tab to selection and hit ENTER)
12. Verify rating specific text appears above text area
13. Enter in optional feedback text in text area if you want
14. Click "Send Feedback"
15. See "Thank you for you feedback" message with a spinner
16. Wait 5 seconds and widget should auto-close
17. Navigate to Admin -> USERS/Feedback 
18. Verify you see your Feedback entry in the Feedback View
19. Repeat steps 2-6 and Hit ESC key
20. Confirm widget closes
21. (Optional) Repeat steps 2-15 
22. Instead of waiting 5 seconds, click the close button and widget should close

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:

For Feedback Backend Integration
+ [ ]  Frontend (Feedback Modal/Widget) calls the Feedback submission REST API endpoint on feedback submission

For Contextual Feedback 
+ [ ] On TANF/FRA Data Files Successful Upload, Feedback Widget appears in the lower-right above the footer
+ [ ] Widget expands to full form once a rating is selected
+ [ ] Widget matches UX mockups in structure, spacing, and interaction behavior
+ [ ] Form includes:
- Required validation for free-text input (if defined by UX)
- Optional anonymity toggle
- Valid Likert input submission
+ [ ] Widget shows successful submission confirmation message with a spinner
+ [ ] After 5 seconds widget auto-closes or user can manually close it by clicking the "X" (close) button
+ [ ] Graceful error handling for backend failures, timeouts, or network issues
+ [ ] Jest unit tests implemented and passing
+ [ ] Testing checklist executed with no blocking issues
+ [ ] README updated (if new dependencies, env vars, or setup steps are introduced)
+ [ ] [Release digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) entry added with screenshots, behavior summary, and testing steps
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

Commit:
#5072 started implementing Feedback Widget and refactoring some code for more abstraction on comon functionality needed for Modal and Widget elements

Commit:
Merge branch '5071-implement-general-feedback-modal-in-tdp-frontend' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
Merge branch '5071-implement-general-feedback-modal-in-tdp-frontend' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 updated imports

Commit:
#5072 formatting Feedback Widget properly and adding functionality for it to appear based on submission of reports

Commit:
#5072 pushing up code for widget positioning and sizing and some refactoring to add hooks

Commit:
Merge branch '5071-implement-general-feedback-modal-in-tdp-frontend' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 updated code after merge

Commit:
Merge branch '5071-implement-general-feedback-modal-in-tdp-frontend' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 fixed placement on widget and fixed some logic

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 updated frontend to use backend api call

Commit:
#5072 removed unused hook and cleaned up some of the code

Commit:
#5072 updated some UI logic for both the modal and widget components. Got the backend api call working and return a response properly to the frontend

Commit:
#5072 cleaning up the UI sizing and overall look

Commit:
#5072 removed testing code

Commit:
#5072 updating and fixing unit test

Commit:
Merge branch '5071-implement-general-feedback-modal-in-tdp-frontend' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 working on fixing tabbing logic and unit tests for it - down to two failing from 7 now

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 finished up updating test and adding test for code coverage. Cleaned up code

Commit:
#5072 refactored tabbing code and and fixed up some last minute UI sizing and positioning

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 updated widget test coverage

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission

Commit:
#5072 quick fix for a white space expansion below footer due to not constrained properly within the document flow

lhawkman27 commented:
<img width="1540" height="833" alt="Screenshot 2025-07-21 at 2 04 20 PM" src="https://github.com/user-attachments/assets/193fb298-8b7a-473f-aa4b-4669c82ca32a" />
<img width="1492" height="823" alt="Screenshot 2025-07-21 at 2 04 41 PM" src="https://github.com/user-attachments/assets/9f9c8dfc-b451-46f6-ad1c-35853e071a8b" />
<img width="1712" height="845" alt="Screenshot 2025-07-21 at 12 09 24 PM" src="https://github.com/user-attachments/assets/dcf7b1e6-2b45-423e-aa84-6f95c13fa53d" />

lhawkman27 commented:
<img width="1592" height="713" alt="Screenshot 2025-07-21 at 12 09 52 PM" src="https://github.com/user-attachments/assets/cd82aed8-2af4-4052-9062-d265eba7492c" />
<img width="1578" height="728" alt="Screenshot 2025-07-21 at 12 10 03 PM" src="https://github.com/user-attachments/assets/9ca6f05e-adf2-4b9a-96bb-cb3165c7f7e1" />
<img width="1513" height="596" alt="Screenshot 2025-07-21 at 12 10 18 PM" src="https://github.com/user-attachments/assets/19e2d31e-af97-42ec-8473-a61a7f8bbf37" />

Commit:
#5072 updated TANF data types being passed to the widget to stay consisent with header messages

lhawkman27 commented:
<img width="1700" height="859" alt="Screenshot 2025-07-21 at 2 18 52 PM" src="https://github.com/user-attachments/assets/bac906f8-a71b-4847-bda9-6f719c782c1d" />

elipe17 commented on tdrs-frontend/src/components/Reports/Reports.jsx:
```diff
@@ -219,6 +224,14 @@ function Reports() {
     setTouched((currentForm) => ({ ...currentForm, stt: true }))
   }
 
+  const handleOpenWidget = () => {
```
can we wrap this logic into a hook since it used the same way in both reports?

lhawkman27 replied:
I wanted to say NEVER!!!! hahahaha. But I moved it into a hook

elipe17 reviewed (changes requested):
This is looking really good! I took some notes/found some bugs for you to check out.

- One thing I noticed is we don't have any identifier to delineate contextual feedback from general feedback. Then within the contextual feedback we need to delineate between TANF data files and FRA data files. This is an issue with both the backend and frontend. The backend needs to accept extra contextual information to help align the feedback. The frontend is responsible for sending the extra context. I wrote #5197 for this issue.
- Should the modal be pinned to the bottom of the view port like the general feedback modal?
- Checkbox outline covers text
<img width="270" height="120" alt="Screenshot 2025-07-22 at 8 49 25 AM" src="https://github.com/user-attachments/assets/9b20999d-1969-47e7-bf84-ed9badd3247b" />

- I think "thank you/close modal" should stay open based on the length of the request. A 5 second delay makes the app feel slow
- The score received text implies the score is sent before the user clicks submit. Maybe this should be the case since feedback interaction is usually always low. Then we could patch the feedback text?
<img width="281" height="89" alt="Screenshot 2025-07-22 at 8 52 43 AM" src="https://github.com/user-attachments/assets/60152831-58fa-4f09-aa0d-b8cd6f58378b" />

- Bug: On TANF data files page, submit a TANF file, wait for the feedback modal to appear, change program type to SSP, see that modal context is now pointing at SSP instead of TANF and vice versa.

mattcoleanderson commented on tdrs-frontend/src/components/Feedback/Feedback.jsx:
```diff
@@ -49,21 +53,24 @@ function Feedback() {
             isOpen={isModalOpen}
             message={
               <p>
                 Your response has been recorded. If you're encountering an issue
                 you need support to resolve please feel free to email us at{' '}
```
```suggestion
Your response has been recorded. If you're encountering an issue
you need support to resolve, please feel free to email us at{' '}
```

mattcoleanderson replied:
I know this isn't part of this ticket, but I noticed it and feel like this sentence needs a comma.

lhawkman27 replied:
Nice catch!!! I added the comma in.

mattcoleanderson commented:
A few notes:

- When selecting a rating or the `Send anonymously` checkbox, the whole modal flashes a blue focused outline until you release the mouse click.
![2025-07-22 14 26 22](https://github.com/user-attachments/assets/9bda8ccc-7610-4a9e-9245-f25942a51076)
- Should the Modal header be focusable?
<img width="324" height="347" alt="image" src="https://github.com/user-attachments/assets/877794e5-6e40-4f55-8388-1b0d40409b98" />

- After choosing a rating you get a text popup that says "Score has been received". I feel like this is misleading since it isn't received until you hit submit.

Commit:
#5072 dev review code updates

lhawkman27 commented:
@mattcoleanderson good catch on the focus bug for the widget. I ran into the with the Feedback Modal in the last ticket. Got that fixed. And I removed the header to being focused on when you click on it.

lhawkman27 commented:
@elipe17 thanks for making that issue. I was thinking while working on this that it would be nice to have an identifier to delineate contextual feedback from general feedback. 

- Fixed the checkbox spacing so text isn't covered by it when selected
- Changed how long the "thank you for the feedback" stays open, and made it so it closes when we get the response back from the api call

* still working on heading title staying consistent with the data file you uploaded

Commit:
#5072 updated unit test

jtimpe commented on tdrs-frontend/src/components/Feedback/FeedbackForm.jsx:
```diff
@@ -0,0 +1,307 @@
+    setIsAnonymous(false)
+  }
+
+  const handleSubmit = useCallback(async () => {
```
curious about the use of `useCallback` here. if i'm understanding [the docs](https://react.dev/reference/react/useCallback) correctly, you're memoizing the function based on the form values. but why?

lhawkman27 replied:
Cause the annoying re-render issue, since I need to call it in the useEffect below it. With the useCallback it avoids re-creating the function every time a re-render happens

mattcoleanderson reviewed (approved):
Thanks for making those changes. This looks good to me. Great work

jtimpe reviewed (approved):
Agree with the other comments on the "score received" language. Other than that, this looks fantastic. Love the hooks for the feedback widget and focus trap. Good work! 👏

Commit:
#5072 fixed header bug found

lhawkman27 commented:
@elipe17 fixed the bug you found. Good find!!!!

Commit:
#5072 removed duplicate variable

Commit:
#5072 added a little more test for widget

elipe17 reviewed (approved):
This looks great sir. I think we just need to update the "score has received" text. Otherwise, LGTM!

Commit:
#5072 changed the likert scale text on selection

lhawkman27 commented:
@reitermb and @kristen-shaw4 
UI/UX Notes/Questions:

1. Likert Scale rating selection text - 
A lot of the developers had questions about the "score has been received" text appearing when you select a rating. It is a bit misleading due to the user feedback not actually being submitted, until the user clicks the "Send Feedback" button.

So I changed the Likert text on rating selection to be:
- GREAT - Fantastic. Would you like to leave a comment?
- GOOD - Great. Would you like to leave a comment?
- FAIR - Thanks. Any suggestions?
- BAD - We're sorry. How can we improve?
- POOR - We're sorry. How can we improve?

Let me know if you like this idea or if you want me to change the text on selection to something else. 

2. Contextual Widget/Modal placement -
A couple of the developers had questions concerning the placement of the widget/modal on the page. One idea was to move the widget/modal all the way to the right corner above the footer. The only issue I can see is the fact that the "General Feedback" sticky button at the bottom of the viewport will overlap the widget, if it is open and the user scrolls up to it. 

If we do want to move the widget/modal all the way over to the right, a fix could be to hide the "General Feedback" button when it touches the widget/modal due to it being open and user scrolling, and then have it reappear when it doesn't. Or another idea is to hide the "General Feedback" button while the Contextual widget/modal is open. This is assuming a user wouldn't want to submit general feedback after they submitted data files and the widget is open for them to submit feedback about their data submission experience. 

What's your opinions on this?

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission

kristen-shaw4 commented:
@lhawkman27 Good point on the language, and it looks good to me with that update.

I was also curious about the placement of the modal since it's higher up on the page than the Give Feedback button and wondered if they either a) should be next to each other? or b) do we need to show the Give Feedback button at all for the general feedback widget if we are showing the feature-specific feedback modal? I am leaning towards b, but open to discussion about it. Something I'd like to consider for the placement is keeping near-ish the submit so that it's still in the user's "line of sight" and noticeable to them when it appears, but at the same time not in their way.

Commit:
#5072 ui review updates. moved widget placement to be where general feedback button is, hides button when data files are submitted and widget is rendered, removed portal and hook, utilizing redux state for opening and closing widget, refactored code, updated unit tests

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission

lhawkman27 commented:
Updated code per UI Review 

- Contextual Feedback Modal/Widget is now positioned where the "Give Feedback" button is at the bottom of the viewport
- "Give Feedback" button is removed and feedback widget is rendered when data files are submitted
- "Give Feedback button shows back up once feedback widget is closed 
- Updated "Thank for feedback widget w/ spinner" to have a 3 sec delay now after we get a response for our request (1 sec delay seemed to be way too fast)
- Changes made solve z-index bug @kristen-shaw4 found when opening the "General" feedback modal after files are submitted

lhawkman27 commented:
<img width="1713" height="849" alt="Screenshot 2025-07-28 at 5 18 14 PM" src="https://github.com/user-attachments/assets/e924aff8-225d-42aa-82ac-e5d420ceabaa" />

lhawkman27 commented:
<img width="1752" height="840" alt="Screenshot 2025-07-28 at 5 19 13 PM" src="https://github.com/user-attachments/assets/ec2d1a19-caa5-48f6-82bb-4914b85635be" />

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission

kristen-shaw4 reviewed (approved):
Looks great! I like that it's persistent when moving around within the page (changing FY & FQ, going between current submission & submission history). 
<img width="320" height="174" alt="Screenshot 2025-07-29 at 2 50 41 PM" src="https://github.com/user-attachments/assets/a42735f5-dd5f-453b-8f71-25970f6d8ab5" />
<img width="477" height="172" alt="Screenshot 2025-07-29 at 3 12 07 PM" src="https://github.com/user-attachments/assets/d175d3c5-3955-42f7-bafc-7096e10d6923" />

Commit:
Merge branch 'develop' into 5072-implement-contextual-feedback-modal-on-data-file-submission
